File: thain-server/src/main/java/com/xiaomi/thain/server/service/FlowService.java
/*
 * Copyright (c) 2019, Xiaomi, Inc.  All rights reserved.
 * This source code is licensed under the Apache License Version 2.0, which
 * can be found in the LICENSE file in the root directory of this source tree.
 */

package com.xiaomi.thain.server.service;

import com.xiaomi.thain.common.exception.ThainException;
import com.xiaomi.thain.common.model.FlowModel;
import com.xiaomi.thain.common.model.JobModel;
import com.xiaomi.thain.common.model.rq.AddFlowRq;
import com.xiaomi.thain.server.model.sp.FlowListSp;
import lombok.NonNull;
import org.quartz.SchedulerException;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.text.ParseException;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;

/**
 * @author liangyongrui
 */
@Service
public interface FlowService {

    List<FlowModel> getFlowList(@NonNull FlowListSp flowListSp);

    Long getFlowListCount(@NonNull FlowListSp flowListSp);

    /**
     * 创建或更新任务
     */
    long add(@NonNull AddFlowRq addFlowRq, @NonNull List<JobModel> jobModelList, @NonNull String appId)
            throws ThainException, ParseException, SchedulerException;

    /**
     * 删除
     */
    boolean delete(long flowId) throws SchedulerException;

    /**
     * 立即执行一次, 返回flow execution id
     */
    long start(long flowId) throws ThainException;

    FlowModel getFlow(long flowId);

    List<JobModel> getJobModelList(long flowId);

    Map<String, String> getComponentDefineStringMap();

    void scheduling(long flowId) throws ThainException, SchedulerException, IOException;

    void pause(long flowId) throws ThainException;

    void updateCron(long flowId, @Nullable String cron) throws ThainException, ParseException, SchedulerException, IOException;
}

File: thain-server/src/main/kotlin/com/xiaomi/thain/server/service/FlowService.kt
package com.xiaomi.thain.server.service

import com.xiaomi.thain.common.exception.ThainException
import com.xiaomi.thain.common.exception.ThainRepeatExecutionException
import com.xiaomi.thain.common.model.ComponentDefine
import com.xiaomi.thain.core.model.dr.FlowDr
import com.xiaomi.thain.core.model.dr.JobDr
import com.xiaomi.thain.core.model.rq.AddFlowAndJobsRq
import com.xiaomi.thain.core.model.rq.AddFlowRq
import com.xiaomi.thain.core.model.rq.AddJobRq
import com.xiaomi.thain.core.model.rq.UpdateFlowRq
import com.xiaomi.thain.common.utils.ifNull
import com.xiaomi.thain.core.ThainFacade
import com.xiaomi.thain.server.dao.FlowDao
import com.xiaomi.thain.server.model.sp.FlowListSp
import org.quartz.SchedulerException
import org.springframework.stereotype.Service
import java.io.IOException
import java.text.ParseException

/**
 * @author liangyongrui
 */
@Service
class FlowService(
        private val flowDao: FlowDao,
        private val thainFacade: ThainFacade) {

    fun getFlowList(flowListSp: FlowListSp): List<FlowDr> {
        return flowDao.getFlowList(flowListSp)
    }

    fun getFlowListCount(flowListSp: FlowListSp): Long {
        return flowDao.getFlowListCount(flowListSp)
    }

    fun add(addFlowRq: AddFlowRq, addJobRqList: List<AddJobRq>, appId: String): Long {
        val flow = addFlowRq
                .takeIf { !it.slaKill || it.slaDuration == 0L }
                ?.copy(slaKill = true, slaDuration = 3L * 60 * 60)
                .ifNull { addFlowRq }.copy(createAppId = appId)
        val flowId = flow.id
        if (flowId != null && flowDao.flowExist(flowId)) {
            val updateFlowRq = UpdateFlowRq(flow, flowId)
            thainFacade.updateFlow(updateFlowRq, addJobRqList)
            return updateFlowRq.id
        }
        return thainFacade.addFlow(AddFlowAndJobsRq(flow, addJobRqList))
                .also { flowDao.updateAppId(it, appId) }
    }

    @Throws(SchedulerException::class)
    fun delete(flowId: Long, appId: String, username: String): Boolean {
        thainFacade.deleteFlow(flowId, appId, username)
        return true
    }

    fun start(flowId: Long, variables: Map<String, String>, appId: String, username: String): Long {
        return thainFacade.startFlow(flowId, variables, appId, username)
    }

    fun getFlow(flowId: Long): FlowDr? {
        return flowDao.getFlow(flowId)
    }

    fun getJobModelList(flowId: Long): List<JobDr> {
        return flowDao.getJobModelList(flowId)
    }

    fun getComponentDefine(): List<ComponentDefine> {
        return thainFacade.componentService.componentDefineList
    }

    @Throws(ThainException::class, SchedulerException::class, IOException::class)
    fun scheduling(flowId: Long, appId: String, username: String) {
        thainFacade.schedulingFlow(flowId, appId, username)
    }

    @Throws(ThainException::class, ParseException::class, SchedulerException::class, IOException::class)
    fun updateCron(flowId: Long, cron: String?) {
        thainFacade.updateCron(flowId, cron)
    }

    @Throws(ThainException::class)
    fun pause(flowId: Long, appId: String, username: String) {
        thainFacade.pauseFlow(flowId, appId, username, false)
    }

}
